### PR TITLE
fix buffer overrun in validate.cpp

### LINKED ISSUE #58: buffer overrun in validate test
reported by Ching Chu:

$ ./validate
buffer size: 327680
off: 0
check 1: MMIO CPU initialization + read back via cuMemcpy D->H
check 2: gdr_copy_to_bar() + read back via cuMemcpy D->H
check 3: gdr_copy_to_bar() + read back via gdr_copy_from_bar()
check 4: gdr_copy_to_bar() + read back via gdr_copy_from_bar() + 5 dwords offset
check 5: gdr_copy_to_bar() + read back via gdr_copy_from_bar() + 11 bytes offset
[1]    316576 segmentation fault  ./validate

### PR DESCRIPTION
**Pull request by drossetti**
while there, fortify and beatify compare_buf()

fixes #58

### CODE REVIEW

File: common.hpp [hunk 1]
```diff
@@ -56,13 +56,20 @@
...
-			++diff;
-		}
+        if (ref_buf[w] != buf[w]) {
+            if (!diff) {
+                printf("%10.10s %8.8s %8.8s\n", "word", "content", "expected");
+            }
+            if (diff < 10) {
+                printf("%10d %08x %08x\n", w, buf[w], ref_buf[w]);
+            }
+            ++diff;
```
Comment (pakmarkthub) [inline common.hpp]: Should report how many diffs are detected from how many elements (i.e. report "diff / w").

File: validate.cpp [hunk 2]
```diff
@@ -110,11 +110,12 @@ int main(int argc, char *argv[])
...
 
+        // check for access to misaligned address
         extra_off = 11;
         printf("check 5: gdr_copy_to_bar() + read back via gdr_copy_from_bar() + %d bytes offset\n", extra_off);
-        gdr_copy_to_bar((char*)buf_ptr + extra_off, init_buf, size);
-        gdr_copy_from_bar(copy_buf, (char*)buf_ptr + extra_off, size);
-        compare_buf(init_buf, copy_buf, size);
+        gdr_copy_to_bar((char*)buf_ptr + extra_off, init_buf, size - extra_off);
+        gdr_copy_from_bar(copy_buf, (char*)buf_ptr + extra_off, size - extra_off);
+        compare_buf(init_buf, copy_buf, size - extra_off);
```
Comment (pakmarkthub) [inline validate.cpp]: Report as failure if any elements are different. Should do this for the other sub tests as well.

**Comment by drossetti** [comment]
@pakmarkthub please have a look

**Comment by drossetti** [comment]
thanks

**Pull request merged**